api/src/main/java/org/apache/iceberg/metrics/FixedReservoirHistogram.java

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * A {@link Histogram} implementation that uses a fixed size reservoir like circular buffer. Oldest
+ * observation is replaced by latest observation.
+ */
+public class FixedReservoirHistogram implements Histogram {
+  private final long[] measurements;
+  private long count;
+
+  public FixedReservoirHistogram(int reservoirSize) {
+    this.measurements = new long[reservoirSize];
+    this.count = 0L;
+  }
+
+  @Override
+  public synchronized long count() {
+    return count;
+  }
+
+  @Override
+  public synchronized void update(long value) {
+    int index = (int) (count % measurements.length);
+    measurements[index] = value;
+    count++;
+  }
+
+  @Override
+  public Statistics statistics() {
+    int size = measurements.length;
+    if (count < (long) measurements.length) {
+      size = (int) count;
+    }
+
+    long[] values = new long[size];
+    for (int i = 0; i < values.length; ++i) {
+      synchronized (this) {
+        values[i] = measurements[i];
+      }
+    }
+
+    return new UniformWeightStatistics(values);
+  }
+
+  private static class UniformWeightStatistics implements Statistics {
+    private final long[] values;
+
+    private UniformWeightStatistics(long[] values) {
+      // since the input values is already a copied array,
+      // there is no need to copy again.
+      this.values = values;
+      Arrays.sort(this.values);
+    }
+
+    @Override
+    public long size() {
+      return values.length;
+    }
+
+    @Override
+    public double mean() {
+      if (values.length == 0) {
+        return 0.0;
+      } else {
+        double sum = 0.0;
+        for (int i = 0; i < values.length; ++i) {
+          sum += values[i];
+        }
+
+        return sum / values.length;
+      }
+    }
+
+    @Override
+    public double stdDev() {
+      if (this.values.length <= 1) {
+        return 0.0;
+      } else {
+        double mean = mean();
+        double sumOfDiffSquares = 0.0;
+        for (int i = 0; i < values.length; ++i) {
+          double diff = values[i] - mean;
+          sumOfDiffSquares += diff * diff;
+        }
+
+        double variance = sumOfDiffSquares / values.length;
+        return Math.sqrt(variance);
+      }
+    }
+
+    @Override
+    public long max() {
+      return values.length == 0 ? 0L : values[values.length - 1];
+    }
+
+    @Override
+    public long min() {
+      return values.length == 0 ? 0L : values[0];
+    }
+
+    @Override
+    public double percentile(double percentile) {
+      Preconditions.checkArgument(
+          !Double.isNaN(percentile) && percentile >= 0.0 && percentile <= 1.0,
+          "Percentile point cannot be outside the range of [0.0 - 1.0]: %s",
+          percentile);
+      if (values.length == 0) {
+        return 0.0;
+      } else {
+        double position = percentile * values.length;
+        int index = (int) position;
+        if (index < 1) {
+          return values[0];
+        } else if (index >= values.length) {
```

Comment:
I think this is correct.

The only time that we would expect `index` to be `equal to `values.length` is when `percentile` is `1.0`, but I think it's reasonable to be extra safe here and check the actual index value produced to avoid an `ArrayIndexOutOfBoundsException`.